### PR TITLE
[frontend] refactor ASTBuilder for binary expression, fix issue#38

### LINKED ISSUE #38: Maybe operator precedence errors
# Cause
https://github.com/Enna1/remniw/blame/709c4c7fa0624a093ca12de565ef38aaa646c556/grammar/Remniw.g4#L44

# Reproduce
## source code
```
func main() int {
  var x int;
  x = 5 - 9 + 4;
  return 0;
}
```
## emited ir
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"

@fmtstr = private unnamed_addr constant [5 x i8] c"%lli\00", align 1
@fmtstr.1 = private unnamed_addr constant [6 x i8] c"%lli\0A\00", align 1

define i64 @main() {
entry:
  %x = alloca i64, align 8
  store i64 -8, ptr %x, align 8
  ret i64 0
}
```
it should be 0 rather than -8. This is because `9 + 4` will be matched before `5 - 9` caused by the precedence of the grammar...
```
   | expr ('+' | '-')  expr # AddExpr
```
this may solve the problem.
~~I guess `*` and `/` have the same problem.~~

### PR DESCRIPTION
**Pull request by Enna1**

### CODE REVIEW

**Comment by Enna1** [comment]
Hi @Sober7135, can you check if this fix https://github.com/Enna1/remniw/issues/38 ?
If no any other comment, I will merge this PR and close https://github.com/Enna1/remniw/issues/38.

File: test/codegen/precedence.rw [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+// Emit RISCV assembly
+// RUN: %remniw-llc --target=riscv %t1 -o %t2.riscv.s ; \
+// RUN:     %riscv-cc %t2.riscv.s -o %t3.riscv.exe; \
+// RUN:     qemu-riscv64 %t3.riscv.exe | FileCheck %s
+
+func main() int {
+    var x, y int;
+    x = 5 - 9 + 4;
+    y = 5-3*3+4;
```
Comment (Sober7135) [inline test/codegen/precedence.rw:16]: A little suggestion: maybe use something like `y = 5 * 5 / 5;` or `y = 5 / 5 * 5;`to check the precedence of `*` and `/`. But not necessary...

**Comment by Sober7135** [comment]
Well done!!!
BTW, thanks to you and your program❤️, I've learned a lot from your program!

**Comment by Enna1** [comment]
Glad to see this repo helps someone.
If you find any bug further, feel free to open a issue or send a pull request.

**Pull request merged**